### 1.除了panic引发中断性错误，还可返回error类型错误对象来表示函数调用状态
```go
type error interface{
    Error() string
}
```

- 标准库 errors.New 和 fmt.Errorf 函数用于创建实现 error 接口的错误对象。
- 通过判断错误对象实例来确定具体错误类型。
```go
package main

import (
	"errors"
	"fmt"
)

// 一个error对象
var ErrDivByZero = errors.New("division by zero")


func div(x, y int)(int, error){
	if y == 0 {
		return 0, ErrDivByZero
	}
	return x / y, nil
}

func main(){
	defer func() {
		fmt.Println(recover())
	}()

	// 这个switch牛逼啊，先一个表达式获取值，然后再switch这个值
	switch z, err := div(10, 0); err {
	case nil:
		println(z)
	// 如果case 到ErrDivByZero，则panic
	case ErrDivByZero:
		panic(err)
	}
}
```

### 2.go类似try catch的异常处理
```go
package main

import "fmt"

// interface{}空接口，代表可以接收任何类型的数据
func Try(fun func(), handler func(interface{})){
	// 进入defer
	defer func() {
		// 接收到err
		if err := recover(); err != nil{
			// 调用第二个函数，将err传递过去
			handler(err)
		}
	}()

	// 调用第一个函数,引发恐慌
	fun()
}

func main(){

	Try(func() {
		panic("Test panic")
	}, func(err interface{}) {
		fmt.Println(err)
	})
}
```

### 3.如何区别使用panic和error两种方式
- 惯例是：导致关键流程出现不可修复性错误的使用panic
- 其他使用error